src/json.hpp

```diff
@@ -1397,120 +1397,77 @@ constexpr T static_const<T>::value;
 /// abstract input adapter interface
 struct input_adapter_protocol
 {
-    virtual int get_character() = 0;
-    virtual std::string read(std::size_t offset, std::size_t length) = 0;
+    virtual int get_character() = 0; // returns characters in range [0,255], or eof()
```

Comment:
Should this be `std::char_traits<char>::int_type` instead of `int` so that we know that there will be no type changes later?

Reply:
That would be an acceptable change, I think.  It's typically that type anyway, and being explicit about it keeps things obvious; that all implementation of the adapter class need to be consistent in producing `std::char_traits<char>::int_type`, too.

Reply:
A `using int_type = ...` would make it less verbose then the full thing.